### PR TITLE
Decode with recursion limit

### PR DESCRIPTION
**Pull request by arkpar**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 [package]
 name = "parity-scale-codec"
 description = "SCALE - Simple Concatenating Aggregated Little Endians"
-version = "1.3.0"
+version = "1.3.1"
```
Comment (bkchr) [inline Cargo.toml:4]: We can not release this version. Master has breaking changes. If you want to release this. I propose that you backport this after merge to master. We have a tag for the last release.

File: src/codec.rs [hunk 2]
```diff
@@ -131,6 +131,14 @@ pub trait Input {
...
 		Ok(buf[0])
 	}
+
+	/// Descend into nested structure when decoding
+	fn descend(&mut self) -> Result<(), Error> {
+		Ok(())
+	}
+
+	/// Ascend to previous structure level when decoding.
+	fn ascend(&mut self) { }
```
Comment (bkchr) [inline src/codec.rs:141]:
```suggestion
	fn ascend(&mut self) {}
```

File: src/depth_limit.rs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	/// Decode `Self` with given maximum recursion depth.
+	/// is returned.
+	fn decode_with_depth_limit(limit: u32, input: &[u8]) -> Result<Self, Error>;
+	/// Decode `Self` and consume all of the given input data. If not all data is consumed, an error
+	/// is returned.
+	fn decode_all_with_depth_limit(limit: u32, input: &[u8]) -> Result<Self, Error>;
+}
+
+
+pub struct DepthTrackingInput<'a, I> {
```
Comment (bkchr) [inline src/depth_limit.rs:32]: Missing docs.

File: src/codec.rs [hunk 4]
```diff
@@ -131,6 +131,14 @@ pub trait Input {
 		self.read(&mut buf[..])?;
 		Ok(buf[0])
 	}
+
+	/// Descend into nested structure when decoding
+	fn descend(&mut self) -> Result<(), Error> {
```
Comment (bkchr) [inline src/codec.rs:136]: From the api docs and the usage I don't understand when we enter a nested structure.

If I have a struct as field of a struct, aren't we entering a new nested structure as well?

Comment (arkpar) [reply src/codec.rs:136]: It only makes sense to trace nested references as nested structures are not allowed in rust anyway. I'll update docs.

File: src/codec.rs [hunk 5]
```diff
@@ -131,6 +131,17 @@ pub trait Input {
 		self.read(&mut buf[..])?;
 		Ok(buf[0])
 	}
+
+	/// Descend into nested reference when decoding.
```
Comment (bkchr) [inline src/codec.rs:135]:
```suggestion
	/// Descend into nested heap allocated structure while decoding.
```

File: src/codec.rs [hunk 6]
```diff
@@ -131,6 +131,17 @@ pub trait Input {
...
+	/// This is called when decoding a new refence-based instance,
+	/// such as `Vec` or `Box`
```
Comment (bkchr) [inline src/codec.rs:137]:
```suggestion
	///
	/// This must be called when decoding a heap-based instance,
	/// such as `Vec` or `Box`.
```

File: src/codec.rs [hunk 7]
```diff
@@ -131,6 +131,17 @@ pub trait Input {
...
+	fn descend(&mut self) -> Result<(), Error> {
```
Comment (bkchr) [inline src/codec.rs:138]:
```suggestion
	fn descend_heap(&mut self) -> Result<(), Error> {
```

Also not that good, but only `descend` isn't enough IMHO.

File: src/codec.rs [hunk 8]
```diff
@@ -131,6 +131,17 @@ pub trait Input {
...
+		Ok(())
+	}
+
+	/// Ascend to previous structure level when decoding.
+	/// This is called when decoding reference-based type is finished.
+	fn ascend(&mut self) {}
```
Comment (bkchr) [inline src/codec.rs:144]: Same as above.

File: src/codec.rs [hunk 9]
```diff
@@ -131,6 +131,17 @@ pub trait Input {
...
+
+	/// Descend into nested reference when decoding.
+	/// This is called when decoding a new refence-based instance,
+	/// such as `Vec` or `Box`
+	fn descend(&mut self) -> Result<(), Error> {
+		Ok(())
+	}
+
+	/// Ascend to previous structure level when decoding.
+	/// This is called when decoding reference-based type is finished.
```
Comment (bkchr) [inline src/codec.rs:143]:
```suggestion
	///
	/// This is called when decoding heap-based type is finished.
```

File: src/codec.rs [hunk 10]
```diff
@@ -405,7 +416,10 @@ impl<T, X> Decode for X where
 	X: WrapperTypeDecode<Wrapped=T>,
 {
 	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
-		Ok(T::decode(input)?.into())
+		input.descend()?;
```
Comment (bkchr) [inline src/codec.rs:419]: `WrapperTypeDecode` is not only used for heap based structures.

File: src/depth_limit.rs [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+		if input.input.is_empty() {
+			Ok(res)
+		} else {
+			Err(crate::decode_all::DECODE_ALL_ERR_MSG.into())
+		}
+	}
+
+	fn decode_with_depth_limit(limit: u32, input: &[u8]) -> Result<Self, Error> {
+		let mut input = DepthTrackingInput {
+			input: &mut &input[..],
```
Comment (KiChjang) [inline src/depth_limit.rs:85]: This here creates a new byte slice that's separate from `input`, which would not advance the `input` index after decoding is done.

Comment (arkpar) [reply src/depth_limit.rs:85]: That would require changing the signature to accept a mutable input. If such a use case is required a PR is welcome

**Pull request merged**